### ISSUE TITLE
fbs buildvm ubuntu command failing

### ISSUE THREAD

**Issue by jim-heath** [mherrmann/fbs#296]
OS: Ubuntu 22.04 (also tried 20.04, 18.04 and 16.04)
Python: 3.9.16 (also tried 3.7.16 and 3.8.0)
fbs Pro: 1.1.9
PyInstaller: 5.9.0
PyQt: 5.15.2

I'm using pyenv to manage Python version on my systems. I followed the Manual page: https://build-system.fman.io/manual/ to create a basic program and it runs fine: ```fbs startproject && fbs run```. I can also successfully build the installer with ```fbs installer``` But when I try to prepare it for release, the following command fails:

```
fbs buildvm ubuntu
```

Here is the output:
```
Sending build context to Docker daemon  24.58kB

Step 1/25 : FROM ubuntu:16.04
16.04: Pulling from library/ubuntu
Digest: sha256:1f1a2d56de1d604801a9671f301190704c25d604a416f59e03c04f5c6ffee0d6
Status: Image is up to date for ubuntu:16.04
 ---> b6f507652425
Step 2/25 : ARG requirements
 ---> Using cache
 ---> dc697d981eb4
Step 3/25 : ARG python_version=3.6.12
 ---> Using cache
 ---> 9934d011eb30
Step 4/25 : ARG python_build_deps="make build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev"
 ---> Using cache
 ---> c755f1fb1e4b
Step 5/25 : RUN apt-get update &&     apt-get upgrade -y
 ---> Using cache
 ---> 4943ba34edf8
Step 6/25 : RUN apt-get install -y curl git
 ---> Using cache
 ---> 4b9647266b66
Step 7/25 : ENV PYENV_ROOT /root/.pyenv
 ---> Using cache
 ---> fa9de5fb4a7f
Step 8/25 : ENV PATH $PYENV_ROOT/shims:$PYENV_ROOT/bin:$PATH
 ---> Using cache
 ---> fbbfa7f733df
Step 9/25 : RUN curl https://pyenv.run | bash
 ---> Using cache
 ---> ec6441f5b26b
Step 10/25 : RUN pyenv update
 ---> Using cache
 ---> 6d9210dc350b
Step 11/25 : RUN echo $python_build_deps | xargs apt-get install -y --no-install-recommends
 ---> Using cache
 ---> 655b9faa0391
Step 12/25 : RUN CONFIGURE_OPTS=--enable-shared pyenv install $python_version &&     pyenv global $python_version &&     pyenv rehash
 ---> Using cache
 ---> ac39bdc700a7
Step 13/25 : RUN apt-get install libgl1-mesa-glx -y
 ---> Using cache
 ---> 6bdde107f148
Step 14/25 : RUN apt-get install ruby ruby-dev build-essential -y &&     gem install --no-document fpm
 ---> Using cache
 ---> 94b2f792a94e
Step 15/25 : WORKDIR /root/GPS Package Builder
 ---> Using cache
 ---> b6582fd38e69
Step 16/25 : ADD *.txt /tmp/requirements/
 ---> Using cache
 ---> 441ba21526a2
Step 17/25 : RUN pip install --upgrade pip &&     pip install -r "/tmp/requirements/${requirements}"
 ---> Running in 5f59df901259
Collecting pip
  Downloading https://files.pythonhosted.org/packages/a4/6d/6463d49a933f547439d6b5b98b46af8742cc03ae83543e4d7688c2420f8b/pip-21.3.1-py3-none-any.whl (1.7MB)
Installing collected packages: pip
  Found existing installation: pip 18.1
    Uninstalling pip-18.1:
      Successfully uninstalled pip-18.1
Successfully installed pip-21.3.1
Collecting fbs==1.1.7
  Downloading fbs-1.1.7.tar.gz (202 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Collecting PyQt6==6.1.1
  Downloading PyQt6-6.1.1.tar.gz (949 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: /root/.pyenv/versions/3.6.12/bin/python3.6 /root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmp_wm9f1b8
       cwd: /tmp/pip-install-23c0lr9p/pyqt6_f9bf8a2281bb4d96bda5655059202f89
  Complete output (29 lines):
  Traceback (most recent call last):
    File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
      hook = backend.prepare_metadata_for_build_wheel
  AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
      main()
    File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 160, in prepare_metadata_for_build_wheel
      whl_basename = backend.build_wheel(metadata_directory, config_settings)
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/sipbuild/api.py", line 51, in build_wheel
      project = AbstractProject.bootstrap('pep517')
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/sipbuild/abstract_project.py", line 83, in bootstrap
      project.setup(pyproject, tool, tool_description)
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/sipbuild/project.py", line 594, in setup
      self.apply_user_defaults(tool)
    File "project.py", line 60, in apply_user_defaults
      super().apply_user_defaults(tool)
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
      super().apply_user_defaults(tool)
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/sipbuild/project.py", line 241, in apply_user_defaults
      self.builder.apply_user_defaults(tool)
    File "/tmp/pip-build-env-hc05f6uw/overlay/lib/python3.6/site-packages/pyqtbuild/builder.py", line 68, in apply_user_defaults
      "specify a working qmake or add it to PATH")
  sipbuild.pyproject.PyProjectOptionException
  ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/a0/07/0ae4f67768c1150af851572fae287aeaf956ed91b3b650b888a856274ae4/PyQt6-6.1.1.tar.gz#sha256=8775244fa73f94bfe8ae7672b624e2a903a22bc35d7ea42dd830949e2f9e43c7 (from https://pypi.org/simple/pyqt6/) (requires-python:>=3.6.1). Command errored out with exit status 1: /root/.pyenv/versions/3.6.12/bin/python3.6 /root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /tmp/tmp_wm9f1b8 Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement PyQt6==6.1.1 (from versions: 6.0.0, 6.0.1, 6.0.2, 6.0.3, 6.1.0, 6.1.1, 6.2.0, 6.2.1, 6.2.2, 6.2.3, 6.3.0, 6.3.1, 6.4.0, 6.4.1, 6.4.2)
ERROR: No matching distribution found for PyQt6==6.1.1

The command '/bin/sh -c pip install --upgrade pip &&     pip install -r "/tmp/requirements/${requirements}"' returned a non-zero code: 1

```

Is there a recommended development environment for Ubuntu (OS version and Python version) so that I don't run into errors?

Thanks,
Jim

**Comment by mherrmann** [comment]
Does this help?

https://github.com/mherrmann/fbs/issues/294#issuecomment-1483955674

**Comment by jim-heath** [comment]
> Does this help?
> 
> [#294 (comment)](https://github.com/mherrmann/fbs/issues/294#issuecomment-1483955674)

Thanks for this tip, that got me further along with running the ```fbs buildvm ubuntu``` command. The build process then complained about not being able to find other packages that my requirements file referenced. I had to uninstall all packages with pip and install older versions listed in the error output. Once I did that the build finished.

Unfortunately when I ran: ```fbs runvm ubuntu``` and was in the Docker container, the ```fbs freeze``` command throws this error:

```
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.6.12/bin/fbs", line 8, in <module>
    sys.exit(_main())
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/__main__.py", line 17, in _main
    fbs.cmdline.main()
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/cmdline.py", line 32, in main
    fn(*args)
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/builtin_commands/__init__.py", line 164, in freeze
    freeze_ubuntu(debug=debug)
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/freeze/ubuntu.py", line 4, in freeze_ubuntu
    freeze_linux(debug)
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/freeze/linux.py", line 8, in freeze_linux
    run_pyinstaller(debug=debug)
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/site-packages/fbs/freeze/__init__.py", line 48, in run_pyinstaller
    cp = run(args, capture_output=True, text=True)
  File "/root/.pyenv/versions/3.6.12/lib/python3.6/subprocess.py", line 423, in run
    with Popen(*popenargs, **kwargs) as process:
TypeError: __init__() got an unexpected keyword argument 'capture_output'
```

I attempted to modify the Docker file as described in [#294 comment](https://github.com/mherrmann/fbs/issues/294#issuecomment-1493125728) but it resulted in the same error as above. It seems the build process ignored my change to the Docker file to use python version 3.7.0 instead of 3.6.12.

What is the ideal development environment to fbs pro working properly?

**Comment by mherrmann** [comment]
I fixed these issues now in releases [1.2.0](https://github.com/mherrmann/fbs/releases/tag/v1.2.0) and [1.2.1](https://github.com/mherrmann/fbs/releases/tag/v1.2.1). If your purchase was within the last 12 months and you let me know the email you purchased with, then I'll give you access to the new version.
